### PR TITLE
Add edgeql to go code generator

### PR DESCRIPTION
**Pull request by fmoor**
This change implements the default behavior outlined in https://github.com/edgedb/edgedb/discussions/4244. All arguments and options to the generator command are left for future work along with query hashing.

The [documentation](https://github.com/edgedb/edgedb-go/pull/236/files#diff-43b8e20634ba03cc187345a9c8ac57669ba567a358d5cfe83e2ed86325362bc8), [help text](https://github.com/edgedb/edgedb-go/pull/236/files#diff-7df6c5ada0b868390ac839ba9e93f37fd6c19a264cdcf317a33561167cea9494R54), and `cmd/edgeql-go/testdata` directory may be areas of interest for reviewers. 

This is related to https://github.com/edgedb/edgedb-go/issues/182

### CODE REVIEW

**Comment by fmoor** [comment]
Is `edgeql-go` an appropriate name for the go codegen binary?

cc @1st1 @colinhacks

**Comment by fmoor** [comment]
Needs tests still

**Comment by fmoor** [comment]
Ready for review

**Comment by fmoor** [comment]
@divan and anyone else that would like, we would love to have early feedback from the community on this feature if you are interested in trying it out.

**Comment by fmoor** [comment]
Currently all generated query functions are public. I think they should be private by default. Then if people ask for it we can provide a mechanism to tell the generator to make query functions public on a per query basis.

File: cmd/edgeql-go/testdata/test-project2/dbschema/migrations/00001.edgeql [hunk 1]
```diff
@@ -0,0 +1,12 @@
+CREATE MIGRATION m1fqtauhtvc2w56wh2676x5g26aye22ghx7b7mtnbfekxpmxmnjx2a
```
Comment (fantix) [inline cmd/edgeql-go/testdata/test-project2/dbschema/migrations/00001.edgeql:1]: This migration file is not used? We don't have the `.esdl` files either.

File: cmd/edgeql-go/gen.go [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+	)
+
+	switch desc.Type {
+	case descriptor.Set, descriptor.Array:
+		imports, err = g.generateSlice(&buf, desc)
+	case descriptor.Object, descriptor.NamedTuple:
+		imports, err = g.generateObject(&buf, desc, required)
+	case descriptor.Tuple:
+		imports, err = g.generateTuple(&buf, desc, required)
+	case descriptor.BaseScalar, descriptor.Scalar, descriptor.Enum:
```
Comment (fantix) [inline cmd/edgeql-go/gen.go:89]: `Scalar` and `Enum` have non-built-in type descriptor IDs, `generateBaseScalar()` will generate a broken Go file currently.

**Review by fantix — COMMENTED** [review]
Pretty cool overall! The anonymous "inline structs" in the generated code saved a lot of naming trouble 👍

File: cmd/edgeql-go/testdata/test-project2/dbschema/migrations/00001.edgeql (hunk 1 above)
Comment (fmoor) [reply cmd/edgeql-go/testdata/test-project2/dbschema/migrations/00001.edgeql:1]: It is here to be sure that it is ignored by the code generator. If a `.go` file is generated and there is not matching `.go.assert` file then a test will fail.

File: errors_gen.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
+// This source file is part of the EdgeDB open source project.
+//
+// Copyright 2020-present EdgeDB Inc. and the EdgeDB authors.
```
[the remark this answers is not in the document]
Comment (fmoor) [reply errors_gen.go:3]: If that is the case I have been doing it wrong for a while :/

File: cmd/edgeql-go/testdata/test-project2/dbschema/migrations/00001.edgeql (hunk 1 above)
Comment (fantix) [reply cmd/edgeql-go/testdata/test-project2/dbschema/migrations/00001.edgeql:1]: Ah makes a lot of sense! 👍

File: cmd/edgeql-go/gen.go (hunk 2 above)
Comment (fmoor) [reply cmd/edgeql-go/gen.go:89]: fixed

**Review by fantix — APPROVED** [review]

File: errors_gen.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (elprans) [reply errors_gen.go:3]: The industry seems to be moving away from including years in copyright headers (e.g React etc), so drop it altogether.

[the remark this answers is not in the document]
Comment (fmoor) [reply errors_gen.go]: So then it should look like this @elprans?

```go
// Copyright EdgeDB Inc. and the EdgeDB authors.
```

**Comment by fmoor** [comment]
TODO: I need to look at linked property support

**Pull request merged**